### ISSUE TITLE
Not authorized error when selecting account from whitelisted domain on iOS devices

### ISSUE THREAD

**Issue by madcatsu** [thomseddon/traefik-forward-auth#185]
**Description:**

After selecting an account from the OAuth accounts list, the error message "Not authorized" appears. After waiting a period of time and refreshing the request to the destination, the OAuth signin request appears again and selecting the correct account again results in the "Not authorized" error. Clearing cookies temporarily resolves the issue until the cookie expires.

**Environment:**

Platform:
- Traefik v2.2
- Docker 19.03.12
- Traefik-forward-auth v2.2.0

Client:

iOS Safari 14.0

**Expected results:** Authentication should pass the user on to the subsequent destination after successful authentication

**Actual results:**
```
time="2020-09-22T11:00:07+08:00" level=warning msg="Error validating csrf cookie"
csrf_cookie="_forward_auth_csrf=234c136effd8b91715bc004df8e1e372"
error="CSRF cookie does not match state" handler=AuthCallback host=oauth.docs.example.com
method=GET proto=https rule=default source_ip=<redacted>
uri="/_oauth?state=e8476dab5591025c0e41d389a3499d65:google:https://docs.example.com/&code=4/4QFxc4Do_EylBxBp9LYuh2Y7Xlq5B2kxMcmK0w_a5iTQB_dPHLnzxmrVSCK-0IT4yquBgBAj0JrfcX2sHA4BBZM&scope=email%20profile%20https://www.googleapis.com/auth/userinfo.profile%20https://www.googleapis.com/auth/userinfo.email%20openid&authuser=1&hd=example.com&prompt=none"
```

Example Docker Compose:
```yaml
traefik:
    container_name: "traefik"
    image: traefik:chevrotin
    networks:
      - public
      - default
    ports:
      - "80:80"
      - "443:443"
    volumes:
      - /etc/localtime:/etc/localtime:ro
      - /var/run/docker.sock:/var/run/docker.sock
      - /opt/appdata/docker/stacks/containerdata/traefik/config:/etc/traefik
      - /opt/appdata/docker/stacks/containerdata/traefik/acme:/etc/acme
    environment:
      DNSIMPLE_BASE_URL: https://api.dnsimple.com
    env_file: ./environment/traefik.env
    depends_on:
      - oauth
      - web
    init: true
    restart: unless-stopped

  oauth:
    container_name: "oauth"
    image: thomseddon/traefik-forward-auth:v2.2.0
    restart: unless-stopped
    networks:
        - public
    volumes:
      - /etc/localtime:/etc/localtime:ro
    env_file:
      - ./environment/oauth.env
    labels:
      - "traefik.enable=true"
      - "traefik.docker.network=public"
      - "traefik.http.routers.oauth.rule=Host(`oauth.docs.example.com`)"
      - "traefik.http.routers.oauth.entrypoints=web-secure"
      - "traefik.http.routers.oauth.tls.certresolver=default"
      - "traefik.http.routers.oauth.tls.domains[0].main=oauth.docs.example.com"
      - "traefik.http.routers.oauth.service=oauth"
      - "traefik.http.routers.oauth.middlewares=oAuthSecured@file"
      - "traefik.http.services.oauth.loadbalancer.server.port=4181"

  web:
    container_name: "web"
    image: registry.gitlab.com/madcatsu/web:latest
    restart: unless-stopped
    networks:
      - public
    volumes:
      - /opt/appdata/docker/stacks/containerdata/web/cache:/var/cache/nginx
      - /opt/appdata/docker/stacks/containerdata/web/pid:/var/run
    read_only: true
    labels:
      - "traefik.enable=true"
      - "traefik.docker.network=public"
      - "traefik.http.routers.web.rule=Host(`docs.example.com`)"
      - "traefik.http.routers.web.entrypoints=web-secure"
      - "traefik.http.routers.web.tls.certresolver=default"
      - "traefik.http.routers.web.tls.domains[0].main=docs.example.com"
      - "traefik.http.routers.web.service=web"
      - "traefik.http.routers.web.middlewares=oAuthSecured@file"
      - "traefik.http.services.web.loadbalancer.server.port=80"
      - "traefik.http.services.web.loadbalancer.server.scheme=h2c"

networks:
  default:
    driver: bridge
  public:
    external: true
    name: public
```

Example forward auth env file:
```
LOG_LEVEL=info
AUTH_HOST=oauth.docs.example.com
COOKIE_DOMAIN=docs.example.com
DOMAIN=example.com
PROVIDERS_GOOGLE_CLIENT_ID=<redacted>
PROVIDERS_GOOGLE_CLIENT_SECRET=<redacted>
SECRET=<redacted>
LIFETIME=43200
```

Example Traefik rules:
```yaml
http:
  middlewares:
    httpsRedirect:
      redirectScheme:
        scheme: https
    httpsOptions:
      headers:
        sslRedirect: true
        sslHost: docs.example.com
        stsSeconds: 2592000
        stsIncludeSubdomains: true
        browserXssFilter: true
        contentTypeNosniff: true
        forceSTSHeader: true
        frameDeny: true
        referrerPolicy: same-origin
        contentSecurityPolicy: |
          default-src https://*.docs.example.com;
          script-src https://*.docs.example.com 'unsafe-inline'
        featurePolicy: |
          autoplay 'none';
          camera: 'none';
          display-capture 'none';
          microphone: 'none';
          usb: 'none'
    googleAuth:
      forwardAuth:
        address: "http://oauth:4181"
        trustForwardHeader: true
        authResponseHeaders:
          - "X-Forwarded-User"
    oAuthSecured:
      chain:
        middlewares:
          - httpsOptions
          - googleAuth
  routers:
    globalRedirectSSL:
      rule: "HostRegexp(`{host:.+}`)"
      entryPoints:
        - web
      middlewares:
        - httpsRedirect
      service: "no-op"
  services:
    # noop service, the URL will be never called
    no-op:
      loadBalancer:
        servers:
        - url: "http://127.0.0.1"
```

**Comment by madcatsu** [comment]
May be related to #176 and #113  given that I also have an SPA with Service Worker fronted by Traefik with traefik-forward-auth.

Looking forward to #178 being implemented in 2.3 to see if this resolve the issue.

**Comment by thomseddon** [comment]
Yes I believe that will be the issue, we have people on our team using iOS without issue, but I'll leave this open and check we're good once 2.3 is released

**Comment by madcatsu** [comment]
Resolved by #178 - thanks @mwitkow @thomseddon
